memory/table_data.go

```diff
@@ -0,0 +1,320 @@
+		td.autoIncVal = 1
+	}
+
+	return td
+}
+
+func allColumns(schema sql.PrimaryKeySchema) []int {
+	columns := make([]int, len(schema.Schema))
+	for i := range schema.Schema {
+		columns[i] = i
+	}
+	return columns
+}
+
+func (td *TableData) columnIndexes(colNames []string) ([]int, error) {
+	columns := make([]int, 0, len(colNames))
+
+	for _, name := range colNames {
+		i := td.schema.IndexOf(name, td.tableName)
+		if i == -1 {
+			return nil, errColumnNotFound.New(name)
+		}
+
+		columns = append(columns, i)
+	}
+
+	return columns, nil
+}
+
+func (td *TableData) numRows(ctx *sql.Context) (uint64, error) {
+	var count uint64
+	for _, rows := range td.partitions {
+		count += uint64(len(rows))
+	}
+
+	return count, nil
+}
+
+// throws an error if any two or more rows share the same |cols| values.
+func (td *TableData) errIfDuplicateEntryExist(cols []string, idxName string) error {
+	columnMapping, err := td.columnIndexes(cols)
+	if err != nil {
+		return err
+	}
+	unique := make(map[uint64]struct{})
+	for _, partition := range td.partitions {
+		for _, row := range partition {
+			idxPrefixKey := projectOnRow(columnMapping, row)
+			if hasNulls(idxPrefixKey) {
+				continue
+			}
+			h, err := sql.HashOf(idxPrefixKey)
+			if err != nil {
+				return err
+			}
+			if _, ok := unique[h]; ok {
+				return sql.NewUniqueKeyErr(formatRow(row, columnMapping), false, nil)
+			}
+			unique[h] = struct{}{}
+		}
+	}
+	return nil
+}
+
+func hasNulls(row sql.Row) bool {
+	for _, v := range row {
+		if v == nil {
+			return true
+		}
+	}
+	return false
+}
+
+// getField returns the index and column index with the name given, if it exists, or -1, nil otherwise.
+func (td *TableData) getField(col string) (int, *sql.Column) {
+	i := td.schema.IndexOf(col, td.tableName)
+	if i == -1 {
+		return -1, nil
+	}
+
+	return i, td.schema.Schema[i]
+}
+
+func (td *TableData) generateCheckName() string {
+	i := 1
+Top:
+	for {
+		name := fmt.Sprintf("%s_chk_%d", td.tableName, i)
+		for _, check := range td.checks {
+			if check.Name == name {
+				i++
+				continue Top
+			}
+		}
+		return name
+	}
+}
+
+func (td *TableData) indexColsForTableEditor() ([][]int, [][]uint16) {
+	var uniqIdxCols [][]int
+	var prefixLengths [][]uint16
+	for _, idx := range td.indexes {
+		if !idx.IsUnique() {
+			continue
+		}
+		var colNames []string
+		expressions := idx.(*Index).Exprs
+		for _, exp := range expressions {
+			colNames = append(colNames, exp.(*expression.GetField).Name())
+		}
+		colIdxs, err := td.columnIndexes(colNames)
+		if err != nil {
+			// this means that the column names in this index aren't in the schema, which can happen in the case of a
+			// table rewrite
+			continue
+		}
+		uniqIdxCols = append(uniqIdxCols, colIdxs)
+		prefixLengths = append(prefixLengths, idx.PrefixLengths())
+	}
+	return uniqIdxCols, prefixLengths
+}
+
+// Sorts the rows in the partitions of the table to be in primary key order.
+func (td *TableData) sortRows() {
```

Comment:
do the enginetests stress this enough? or should some of these longer helpers have unit tests?

Reply:
I think so, yes. We do this every time we edit a pk table in the tests.